### ISSUE TITLE
Patchers refactoring

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#312]
Opening this to group patchers related refactorings

- Config phase belongs to the gui side of patchers:
 - [ ] Move the methods/attributes that belong to the config phase in gui_patchers
 - [ ] Rename "text" to patcher_label (duh) and move it along with tip to the gui_patchers
- [ ] Investigate why we instantiate the gui panels _once_ - leaks some mb of memory
- [ ] Docs ! Docs ! And Docs !

**Comment by Utumno** [comment]
Ok I now had to look more carefully into the records classes. The problem to me is that it uses the `__set/getattr__` pattern. I am not a big fan of this side of python cause it completely breaks static analysis. I am sure however that people knew what they were doing back in 2005/6/7 etc - so there should be performance and other such reasons. But nearing 2020 and as we prepare to move to python 3 we should revisit the desing - maybe using some kind of memory map/ cython or what not would make the patcher super faster. Plus all this dynamic set/getattr and setting different attributes based on isisntance checks should be replaced by good old polymorhism and factories. Food for thought:

- `flags` which is all over should be replaced by a MreRecordWithFlags subclass and encapsulate complexity there.
- RecordHeader should be replaced by two different kinds of Header and a factory that takes care of the if check: https://github.com/wrye-bash/wrye-bash/blob/b3c685e7eeea0659593bc0e761db13786267f9ad/Mopy/bash/brec.py#L130-L137

**Comment by Utumno** [comment]
I had an [illumination](https://github.com/wrye-bash/wrye-bash/compare/e6b40c3105e63f82e521bad34cb157f1890f6e27...utumno-312-patchers-broken), probably due to seeing your refactoring attack on the records @Infernio - it's fairly broken but I think it's in the right way - will clean up a bit (so you can review, I think there are a couple of bugs revealed) and it's still broken but that decouples the gui patchers from the data model patchers and I think will greatly help with going forward with the patcher

Basically I was trying to device a common Base class for gui and Data model - well I couldn't and the reason is they shouldn't be MI linked in the first place. Lots of hacks but less than before - and much more typing

**Comment by Utumno** [comment]
I pushed some rename commits on nightly (along with @GandaG ) that are prerequisites for utumno-312-patchers-broken - this one is not so broken but we have to test and smooth out. @Infernio what's the deal with `allowUnloaded`? Seems most CBash patchers have it `True` but is this really correct?

**Comment by Infernio** [comment]
Will build another nightly with your commits.
I'm also getting pretty close to squashing `inf-records-refactoring` and shoving it onto nightly for some testing.

> what's the deal with allowUnloaded? Seems most CBash patchers have it True but is this really correct?

Not a clue, I know my way around PBash records code and the PBash patchers now but I've always been too scared to look at the CBash patchers :/

**Comment by Utumno** [comment]
allowUnloaded permits to "scan" mods that are not active/merged but yes I don't know either what this scanning does

**Comment by Utumno** [comment]
Dug up a bit on allowUnloaded -  I landed on introduction of allowUnloaded:

b585812e1a6093ca46042b16e86c7ff3bb3498f5

It is listed as a change in Pbash (??)

This swicthed default of allowUnloaded to True:
1790cf0d5db85c297ceff015e72eaceb0a4a9a95

See also:
Fix for CBash Cell Patcher, so it can import from unloaded mods
17653d7128d7f9fff797ad4d9d7fedfdf99a6405
and:
CBash Import Road patcher now allows non-active mods so that it is consistent with the workaround required by the non-CBash patcher:
0fec7a017db60557a0cc64f5cfa4441b49230ad3

and some more in 

2fe16b479b6f5001934f382383281b58b8e505f7

5ed88d6e000acfdf6813440d2fdb171509f99235

I still do not quite get what is the logic checking mods that are not active @lojack5 can you shyed any light?

----

Apart than that utumno-312-patchers-broken is not so broken and feedback appreciated - will leave some comments on commits

**Comment by lojack5** [comment]
Real hazy 'cause it's been forever, but I believe `allowUnloaded` allows certain patchers to import changes from deactivated mods?  Some mods are intended never to be activated, but instead just to have their applicable changes imported: The one that comes to mind is the All Natural Weather Filter mod.  I may be misremembering things though, would have to dig through a bunch of code to be sure.

**Comment by Arthmoor** [comment]
Yep, All Natural is one, and my more recent Oblivion DLC Delayers is another that's not meant to be activated. Both are having issues working properly right now because of other problems with building CBash patches.

**Comment by Utumno** [comment]
@Arthmoor are those mods meant to have some special bash tags? What are the CBash issues?

**Comment by Infernio** [comment]
@Utumno On nightly right now, I get this when hovering over patchers in the patch dialog:
```
Traceback (most recent call last):
  File "bash\basher\patcher_dialog.py", line 505, in OnMouse
    self.gTipText.SetLabel(patcherClass.patcher_tip)
AttributeError: 'NamesPatcher' object has no attribute 'patcher_tip'
```
Seems to have been introduced in 2b0486e92af6f7fc157dfa0c199d0fe88e56d1bf?

**Comment by Utumno** [comment]
ooops good catch will fix

**Comment by Infernio** [comment]
@Utumno I had some commits on nightly earlier, but I retracted those again to stick them onto a separate branch. Problem now is that I can't just rebase and drop the commits from nightly, that creates ~50 conflicts due to the patchers refactoring commits on nightly. Mostly the readClasses/writeClasses pruning and the tweak_name stuff are affected.
I could either keep those commits on my branch for now, or we could push them to dev (don't know how ready they are, at least two of those seem like they need squashing?)

**Comment by Utumno** [comment]
> so I'll just take the dive and merge enough of inf-records-refactoring to merge the full inf-461-patchers

Yep do and fish the 0L ones that depend on those you can squash them no need to be kept separate

**Comment by Utumno** [comment]
> due to the patchers refactoring commits on nightly. Mostly the readClasses/writeClasses pruning and the tweak_name stuff are affected.

Yep, I will proceed merging those - I don't have access now - will do ASAP

**Comment by Utumno** [comment]
Re: commits on nightly - I went through but had no time to comment - the one that merged the `%d` stuff it's ok by me although as discussed it was a goal to have the translatable strings free of those format specifiers. Now the real solution to all this would be to actually use f-strings - much more readable. Yet another reason to speed up py3

**Comment by Infernio** [comment]
@Utumno Alright, ~~wasted~~ spent a few fun hours rebasing and resolving conflicts, but I'm finally done.
`inf-records-refactoring` and `inf-461-patchers` are now on dev, alongside all the commits you commented on as well as a few related ones that I couldn't easily drop.
Since #190 is the next step on the journey to #460, I went ahead and rebased that branch and synced the squashed version on nightly with it so that your commits can get some testing. Will build a new nightly tomorrow, but feel free to push any other commits you think are ready to dev in the meantime :P

**Comment by Infernio** [comment]
Finally got around to doing some work on #312. `inf-312-parser-abc` is rebased and fixed up to the point where it's working as intended for the two refactored parsers. Threw in a new, much-needed Bash Tag just to demonstrate the concept :)

It does rely on parts of `inf-480-records-refactoring-pt2` now, so it might make sense to split that issue into three parts - so we can merge pt2 soon as a prerequisite for 312 work, and leave the record header untangling for a pt3.

**Comment by Utumno** [comment]
Sounds like a plan - was about to post some of the fixes I had on nighty squashed on dev gimme a sec

**Comment by Utumno** [comment]
Time we merge pt1 here too - better break something and fix in this implementation than keep hunting bugs in the old one - plus it's yet another blocker. TODOs:

- allowUnloaded - is it moved ok?
- bodyTags: need a game var and check is handled ok by that mysterious patcher

Other than that a bit of squashing and dev I'd say - can you lend a hand with the TODOs?

**Comment by Infernio** [comment]
`allowUnloaded` looks correct to me now, plus the one in `CBash_ImportInventory` looks correct as well (currently has a `# FIXME CORRECT? comments seem to say so` note). Looking at `bodyTags` now.

**Comment by Infernio** [comment]
So, progress update: in the middle of untangling some CBash nonsense in `multitweaks_names`, I noticed that the MGEF indexing PBash does is broken (doesn't cache correctly), so that's getting fixed as well.

**Comment by Infernio** [comment]
@Utumno Just thought I'd let you know that I'm doing some pretty heavy refactoring on the tweakers, so those will conflict badly soon :P
Currently at:
```
28 files changed, 1008 insertions(+), 1404 deletions(-)
```

**Comment by Utumno** [comment]
Great, yeh they clearly needed some dusting, my idea was that most must become static (I mean kill init overrides) and I hated that `key` attribute, that sometimes also serves for the tweak's record sigs

- I fell into a black hole in work - I will try to tidy up merges to come from my part this weekend :)

**Comment by Infernio** [comment]
> I hated that key attribute, that sometimes also serves for the tweak's record sigs

You mean that `self.key` mess in `multitweak_names`? Yeah, that's gone :)

**Comment by Infernio** [comment]
Crappy premature testing with only two tweaks suggests that my tweak rework, which I've termed 'tweak pooling', is quite a bit faster:
```
0:00:50.868 # nightly
0:00:40.711 # inf-312-tweak-pooling
```

**Comment by Utumno** [comment]
As R. Hettinger put it when he was running out of time in that talk: "I'll do efficiency fast: don't copy data unnecessarily" - dots are nothing in comparison it seems :)

**Comment by Infernio** [comment]
Better comparison, the previous one had the Contents Checker and Tweak Settings enabled as well:
```
0:00:27.333 # nightly
0:00:08.678 # inf-312-tweak-pooling
```
Whew, that is a *lot* faster 👀

**Comment by Infernio** [comment]
Always fun when some bugs you never knew existed disappear during refactoring and you only notice when comparing the before and after 😄 
CBash was unable to replace quest stage text when using `Tweak Names` - fixed itself when I unified the PBash/CBash implementations of the affected tweak

**Comment by Utumno** [comment]
Now where would that be?

**Comment by Infernio** [comment]
Will push in a bit, still need to finish up tweak settings and the race tweaks.

**Comment by Infernio** [comment]
Gah, rabbit holes...
Was getting ready to push and tested with all settings tweaks for Skyrim selected, turns out a whole bunch of tweaks from a3a8f555beb71950b9be5ecba3986bb5078e8cde are not in the pickle files 😬

**Comment by Utumno** [comment]
Take your time - meanwhile I am cleaning up nightly

**Comment by Utumno** [comment]
Left some comments on nightly, the bsa's commits are dev ready I think. Also probably the SSS stuff on top of nightly. We need to leave just the game/fomod/records merges in there (and the XXX), I would like to merge this pt1, the rest we can handle interchangeably with strings prefixing

Once you exit the rabbit hole that is :)

**Comment by Infernio** [comment]
Shouldn't have doubted @BeermotorWB, it was just my buggy code :P
Only the race tweaks left to go through now

**Comment by Infernio** [comment]
Randomly stumbled upon this, `List Patch Config` on nightly produces this:
```
Traceback (most recent call last):
  File "bash\balt.py", line 1793, in __Execute
    self.Execute()
  File "bash\basher\mod_links.py", line 1137, in Execute
    _gui_patchers.sort(key=lambda a: groupOrder[a.patcher_instance.group])
  File "bash\basher\mod_links.py", line 1137, in <lambda>
    _gui_patchers.sort(key=lambda a: groupOrder[a.patcher_instance.group])
AttributeError: 'CBash_AliasesPatcher' object has no attribute 'patcher_instance'
```

**Comment by Infernio** [comment]
@Utumno Here's a wiki page explaining tweak pooling: https://github.com/wrye-bash/wrye-bash/wiki/%5Bdev%5D-Tweak-Pooling
Hope that makes it at least somewhat understandable, it's a pretty radical reworking of the tweak backend after all :)

**Comment by Utumno** [comment]
A great read thanks! I had figured out the gist (wow!) now for the details. meanwhile, I keep prefixing - now that you are there make sure this `tweak.key` attribute is renamed :P

**Comment by Infernio** [comment]
I've actually mostly gotten rid of it locally, just the race tweaks (again) left, then all the keys will be static

**Comment by Utumno** [comment]
Going to rebase 312-patchers - wish me luck and patience :P

**Comment by Infernio** [comment]
@Utumno The early `game/*/patcher` import broke the `-P`/`--Python-mode` CLI switch:

`game/oblivion/patcher/__init__.py` imports `game/oblivion/patcher/importers.py`, which imports from `cint.py`. `cint` does this to check if CBash is enabled:

```python
#See if cint is being used by Wrye Bash
from .bolt import CBash as CBashEnabled
```

But it's too early for that - `bolt.CBash` only gets set at the end of `_import_bush_and_set_game`, we're still in the middle of it.

Not exactly sure how to fix that. We can't just move the `bolt.CBash` assignment up since it needs `bush.game.Esp.canCBash`, for which the games need to be initialized first.

**Comment by Utumno** [comment]
Maybe use a function in CBash then?

**Comment by Infernio** [comment]
Had a bit of an epiphany, mabye: I think `_SimpleImporter` should be renamed to `_APreserver`, and we should split `importers.py` into `preservers.py` and `mergers.py`. Preservers are simple, they just forward the last change from a tagged mod. Mergers instead combine changes from multiple mods, generally guided by one or two tags.

There already is this implicit convention in the docs to differentiate between import patchers that 'preserve' and ones that 'merge', based on the word their description in the readme starts with. This would just be making that official, and it nicely coincides exactly with the patchers that we've managed to reduce down to nothing but a `_SimpleImporter`.

The challenge then will be creating a base class similar to `_APreserver` for mergers (`_AMerger`), most likely by refactoring my recent Import Inventory reimplementation. Would first have to figure out a way to add `ForceAdd` support to that (to absorb `Actors.AIPackagesForceAdd` and `Actors.SpellsForceAdd`), then split existing merge tags into an Add/Remove pair (e.g. `Actors.Spells` -> `Actors.Spells.Add` & `Actors.Spells.Remove`) with a backwards compatibility alias (framework for that is implemented in the Invent splitting commit).

**Comment by Utumno** [comment]
👍 

I am away from my windows machine - I have to reword some commits and then we merge 480-pt2?

but this probably has to wait till monday

**Comment by Infernio** [comment]
Ugh, that was a nightmare. `inf-312-tweak-pooling` is rebased. At some point I must have made a mistake, cause that branch had both the brec package and brec.py in it 🤔

**Comment by Utumno** [comment]
Oh yes conflict resolution on deleted file - you should have had a conflict at some point on brec so it added back brec for you to resolve -> solution should be to delete brec and carry the changes to the split files, so you probably did that but did not delete brec. I had a bunch of rebased branches to save you from that nightmare but yes you surely felt you should redo it - I would have done the same. Will be back come Tuesday, will have a look ASAP

**Comment by Utumno** [comment]
Well I went ahead and rebased the inf-312-tweak-pooling-nightly-2-xxx, just so I masochise a bit :P (plus I would need the rerere for pt3 rebase) - there are some tiny differences that might be important - have a look. I also carried some renames from that branch to your branch ( a1127932063960a59849d114a5bdb8b510ea9189 and parent feel free to squash)

**Comment by Infernio** [comment]
Merged tweak pooling, that rebase was the last straw :P
Still a few behavior regressions, but it's functional and clears a gigantic blocker for CBash removal etc.

I left the two `xxx-tweak-pooling-on-nightly` branches around since there's other stuff on there and I'm not sure if you included that in 480-pt3 yet?

**Comment by Utumno** [comment]
Thanks Inf now I am free to work on pt3 ,plus I will spend some quality time admiring the new codebase :D :D :D

Will get to those branches ASAP

**Comment by Infernio** [comment]
I added two more ideas I had today to the todo list here (see recent edits on the issue).

**Comment by Utumno** [comment]
Went ahead and did the same  - some of it is really big changes not for now but it's great we have all our recent discussions there

**Comment by lojack5** [comment]
Just saw the bullet point about `struct.unpack_from`.  Looked into it a little, to be able to use it the streams we use must implement the [Buffer Protocol](https://docs.python.org/3/c-api/buffer.html).  From a few initial tests, it looks like the only standard lib way to get a buffer from an open file *without copying the data* is by opening it as a memory mapped object: [mmap](https://docs.python.org/3/library/mmap.html).  That would also give a performance improvement most likely.

**Comment by Infernio** [comment]
Memory mapping is definitely something we want to look into, can give huge performance improvements.

**Comment by lojack5** [comment]
I did a small test a few weeks ago swapping all of the modfile opens to use mmap, no noticeable improvement.  But maybe we won't see that until we also us unpack_from, since there's so much use of struct in mod reading?

**Comment by Infernio** [comment]
Probably, yes. But I haven't looked into it yet, so I'm pretty much just guessing 🤷‍♀️

**Comment by Infernio** [comment]
Another round of nightly tracebacks:
```
traceback (most recent call last): 
file "bash\gui\events.pyo", 
line 165, in _post file "bash\balt.pyo", 
line 884, in _conversation_wrapper file "bash\basher\patcher_dialog.pyo", 
line 175, in patchexecute file "bash\patcher\patch_files.pyo", 
line 290, in buildpatch file "bash\patcher\patchers\preservers.pyo", 
line 589, in buildpatch file "bash\patcher\patchers\preservers.pyo", 
line 573, in handle patch cellblock file "bash\bolt.pyo", line 1675, 
in floatsequal file "bash\bolt.pyo", 
line 1668, in isclose typeerror: unsupported operand type(s) for -: 'nonetype' and 'nonetype'
```
No idea why this user's is entirely in lowercase 🤷‍♀️

**Comment by Utumno** [comment]
Correct! The _MelNum default so we ere using it as a float - will fix that one

**Comment by Infernio** [comment]
@Utumno New nightly traceback:
```
Traceback (most recent call last):
  File "bash\gui\events.py", line 177, in _post
    result = listener(*listener_args)
  File "bash\balt.py", line 1125, in DoItemMenu
    self.context_links.popup_menu(self, selected)
  File "bash\balt.py", line 1619, in popup_menu
    link.AppendToMenu(to_popup, parent, selection)
  File "bash\balt.py", line 1841, in AppendToMenu
    link.AppendToMenu(subMenu, window, selection)
  File "bash\balt.py", line 1932, in AppendToMenu
    if not self._append(window): return
  File "bash\basher\mod_links.py", line 1716, in _append
    test_parser = self._parser()
  File "bash\basher\mod_links.py", line 2166, in _parser
    return SpellRecords(detailed=self.do_detailed)
  File "bash\parsers.py", line 1452, in __init__
    atts = bush.game.spell_stats_attrs if self._called_from_patcher else \
AttributeError: 'SpellRecords' object has no attribute '_called_from_patcher'
```

**Comment by Utumno** [comment]
@Infernio mainly - have a look at https://github.com/wrye-bash/wrye-bash/compare/1a6eda0428c44f9b707979f59b5c823467cf05ba...afcfb0f9876bdc4aeb629f93a5abbbac84ddb159 if you have some time. It seems all the caching of MGEFs is not really used- we cache by eid (and eid turned to an int?) although we only query by `sig` - can it be possible? I think that apart from that range the first part of this branch ( https://github.com/wrye-bash/wrye-bash/compare/4d77a078d9cf18e21fdd6c17f6d920d28ea355f4...1a6eda0428c44f9b707979f59b5c823467cf05ba ) is stablish and brings much needed decoupling (note the bush imports deep inside the records code gone)

**Comment by Infernio** [comment]
I'm unsure about how exactly it's supposed to work. MGEFs have FourCCs as their EDIDs (also called the 'effect ID', e.g. the 'Fire Damage' effect [has an EDID of 'FIDG'](https://en.m.uesp.net/wiki/Oblivion:Fire_Damage)). The whole int packing thing is probably supposed to help with looking up based on either unicode or bytestring effect IDs?

**Comment by Utumno** [comment]
Yes but I think we never look up cached_school / name / hostile using the
eid (I may we'll have missed something for instance some conversion of
signs to eids). In that case they are unnecessary - but maybe they were
meant to catch added mgefs or obme ones.In particular the test for
record.flag.hostile/obme check in _index_mgef seems not repeated in
is_harmful. OTH this greatly simplifies the tweaks hierarchy and makes it
more extendible (as it's the records that need modification) so let's clean
it up :)

**Comment by Infernio** [comment]
🤷‍♀️ This all predates me, maybe just drop it and see if it breaks on nightly/during a beta?

**Comment by Utumno** [comment]
Still haven't gotten to TempRefs merge records but I have some interesting work on #312 I really enjoyed doing and hope you will like:

https://github.com/wrye-bash/wrye-bash/compare/4fd030b211d20c5a08eab1c4d0d441703822e277...utumno-647

Took all day and I may have missed something (see also todos sprinkled there) but we are closing in load_plugin in the patch and this will be huge. Meanwhile I have an important question I am really foggy about. In `patch_files.PatchFile.scanLoadMods` we load every mod in the load order using either the mergeFactory or the readfactory. The last one contains all signatures any active parser is interested in. Meanwhile in initData we load (for some patchers) their sources (and their masters) but only the `_read_sigs` of the respective patcher. Now, do we need to load _all_ the signatures (of all the patchers) in scanLoadMods ?
The loaded mod is passed through the scanModFile that _again_ looks in the `_read_sigs` only (more or less - hence this WIP refactoring to hunt down _read_sigs and their use in scanModFile) but also from `update_patch_records_from_mod` - does this one need _all_ the active_read_sigs for all patchers? (I am aware that during merging the readFactory is further updated, this looks ok). The idea is we replace init_pacther_factories with a more sophisticated version that collects all the signatures _per patcher_ for each mod - then loads one mod after the other using only those for initData and reuses the cached mod in scanLoadMods.

**Comment by Infernio** [comment]
Fair warning, it's been a while since I dove into the patchers:

scanLoadMods has to load at minimum for a plugin:
 - If the plugin is being merged: All mergeable tops
 - If the plugin is active: All tops that already have records in the BP (so that update_patch_records_from_mod can update all those records in the BP)
 - If the plugin is being imported from: All tops needed by all patchers that respond to the tags on the plugin
 - If any patchers are active that don't use tags (e.g. tweaks, checkers): All tops wanted by those patchers

There may be some I'm forgetting here. It's easier (and safer) to load too many tops, of course, which is why the current code doesn't bother with this level of granularity.

IMO the ideal end goal would be to have "lazily unpacked" tops, so that we can read each plugin once and keep it in memory, unpacking the binary blobs for each top only once it's needed by a patcher. Then we could easily reuse ModFile instances for all initDatas and scanLoadMods, while ditching the 'specify ahead of time what signatures you need or blow up if you get it wrong' behavior.

**Comment by Utumno** [comment]
Thanks I will have to digest this :p - meanwhile yes, lazily loading from the MobBase `grup_blob` could be an way out, but then it's too much memory - of course keeping the cached mods would be lots of memory too but in the tradeoff speed/memory I would like to be a bit more conservative (hence keeping only the needed tops). Another reason I would want that, and a more important one, is actually to _clarify_ the patch procedure more. I will try to double check your list against the code and if it proves feasible I would try and load those sigs and only those - this is super subtle (think also isActive and mergeFactory updating the read/loadFactory and so on) but I'd rather fix that than leave all this in the dark. Architecturally a fair part of the confusion (and bugs) comes from the fact that loadFactory belongs to ModFile -> injected in Mobs (647 realizations, I have greatly simplified and made LF game agnostic) - the patchers having a loadFactory it's wrong design (been there forever of course)
